Fix #2922: Hide hamburger menu >768px, add function to handle navbar overflow

Oishikatta opened the pull request:
This PR hides the hamburger menu on most pages at >768px by changing NORMAL_NAVBAR_CUTOFF_WIDTH_PX from 800px to 768px.

This PR uses the $timeout method to run after the navbar has rendered. Still todo look into using a second navbar offscreen to measure.

Example video: https://drive.google.com/open?id=0ByhzG72M0KJuRkNKSzRpa2VSS2M

Overflow on page load is very brief, overflow during resize is up to about 500ms.

seanlip commented:
@jblair87 @jaredsilver Please could one of you take a look at the UI aspects of this? Once that's done, I'll do a code pass.

Thanks!

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
         $scope.windowIsNarrow = windowDimensionsService.isWindowNarrow();
+        var handleOverflowResizeTimer;
+        var initialWindowWidth = windowDimensionsService.getWidth();
+        var hideNavElements = {
```
The way this is named, it sounds like a verb (which generally indicates a function). Perhaps something like navElementsVisibilityStatus?

Oishikatta replied:
Renamed

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+          if (initialWindowWidth < windowDimensionsService.getWidth()) {
+            $.each(hideNavElements, function(element, state) {
+              if (state === true) {
+                $('[translate=' + element + ']')
```
Can you instead use an ng-show on the element that refers to $scope.navElementsVisibilityStatus? I.e., the Angular way instead of the jQuery way.

Oishikatta replied:
Fixed

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+          // If the window is resized larger, try displaying the hidden elements
+          if (initialWindowWidth < windowDimensionsService.getWidth()) {
+            $.each(hideNavElements, function(element, state) {
+              if (state === true) {
```
`if (state)` should suffice.

Oishikatta replied:
Fixed

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+              }
+            });
+          }
+          initialWindowWidth = windowDimensionsService.getWidth();
```
Maybe call the variable currentWindowWidth, given that it doesn't always represent the initial window width.

Oishikatta replied:
Fixed

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
           $scope.$apply();
           // Close the sidebar, if necessary.
           SidebarStatusService.closeSidebar();
+          // Limit calls to handleOverflow() to one per 500ms to prevent flicker
```
We have a debouncer service in app.js. Maybe use that?

Oishikatta replied:
I tried to use the debouncer service as it's used here: https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/library/SearchBarDirective.js#L156

However when I tested it didn't seem to have any effect: the function is called with no debouncing. When I tested in the Library search bar, it seemed to also have no debouncing.

Can you take a look and see if I'm using the service correctly?

Thanks.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+          handleOverflowResizeTimer = $timeout(handleOverflow, 500);
         });
 
+        var handleOverflow = function() {
```
I find this method hard to understand. Is its aim to truncate the navbar? If so, maybe call it truncateNavbar? A docstring wouldn't go amiss. either.

Oishikatta replied:
Renamed and added docstring. I wasn't sure on the formatting of the docstrings for the javascript files though, so I used JSDoc format. I didn't a docstring format listed in the contributing wiki/frontend doc.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+          // Wait until i18n is completed.
+          $('.oppia-navbar-tabs a[translate], ' +
+            '.oppia-navbar-tabs span[translate]').each(function(i, element) {
+              if (element.innerText.length === 0) {
```
I don't think I understand the point of all this. Could you please explain?

Oishikatta replied:
Added docstring to explain.

This part of the code is checking if the <span> and <a> elements that are to be translated have any text content - on initial page load, before i18n, they don't. After i18n the text content is added.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+
+          var navWidth = 0;
+          $('ul.oppia-navbar-tabs').children().each(function(i, element) {
+            // Some nav elements have invalid widths from floating
```
Add a full stop after each sentence; ditto below.

Also, maybe in this comment, give an example of which nav elements have invalid widths and what their children are? The overall approach is a bit ad hoc so some more explanation is probably good.

seanlip replied:
Also, it might be worth breaking some of the logic out into a separate method like calculateNavbarTabsWidth(). The current method is fairly long/complicated.

Oishikatta replied:
Added explanation of which elements have invalid widths and broke out into a function.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+
+          $('ul.nav.oppia-navbar-tabs').css('min-width', navWidth);
+
+          if ($('div.collapse.navbar-collapse').height() > 60) {
```
This could do with more explanation, too.

Oishikatta replied:
Added explanation.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
+          }
+        };
+
+        // For Chrome, timeout 0 appears to run after i18n.
```
Can you explain the significance of this a bit more?

Oishikatta replied:
I measured the height of the navbar to be 59.xxx when not overflowed (in inspector), so I used a check for >60 in the function.

If it's over 60, I go through each of the elements to hide: if the element is already hidden, go to the next. If the element is not hidden, hide it and then call the handleOverflow *to be renamed function again after 10ms (to give the browser time to render the css change). Only one element will be hidden per call of the function.

seanlip replied:
OK -- can you add this, and other pertinent info, as comments in the code? Then I'll take a pass over it; the aim is to make it understandable to future devs. Thanks!

Oishikatta replied:
Added comments/docstrings.

jblair87 commented:
To be honest, I've never seen navbars  coded this way. I've always used complete css. I would have built two and used display:none; at the break point. 

Get Outlook for iOS




On Mon, Jan 16, 2017 at 6:50 PM -0500, "Sean Lip" <notifications@github.com> wrote:












@seanlip requested changes on this pull request.



(Oops -- forgot to set status as "request changes".)



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

seanlip commented:
I think the reason we're doing it like this is because of i18n -- the navbar text takes up different lengths in different languages. So we need a more dynamic solution to prevent inadvertent overflow.

Oishikatta commented:
@seanlip I've addressed the review comments, it's ready for another look.

One issue in particular that came up is that I wasn't able to use the debouncer service properly: https://github.com/oppia/oppia/pull/2927#discussion_r96792642

Thanks,
Corey

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,114 @@ oppia.directive('topNavigationBar', [function() {
+
+          // If the window is resized larger, try displaying the hidden elements
+          if (currentWindowWidth < windowDimensionsService.getWidth()) {
+            $.each($scope.navElementsVisibilityStatus, function(e, visible) {
```
Prefer not to use jQuery when not needed. Could you do `for (key in $scope.navElementsVisibilityStatus`, etc.?

Oishikatta replied:
jQuery use has been removed.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,114 @@ oppia.directive('topNavigationBar', [function() {
+          var navWidth = 0;
+          $('ul.oppia-navbar-tabs').children().each(function(i, element) {
+            // The "create" button has a 0 width on its top-level element.
+            // The <li> tag from the Donate button has a large/invalid width.
```
This sounds weird. Should we fix it?

Oishikatta replied:
This would be fixed coincidentally by switching to inline-block.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,114 @@ oppia.directive('topNavigationBar', [function() {
+
+          // Measured non-overflowed navbar height under 60px via inspector.
+          if ($('div.collapse.navbar-collapse').height() > 60) {
+            $.each($scope.navElementsVisibilityStatus, function(e, visible) {
```
Ditto re $.each

Oishikatta replied:
jQuery use has been removed.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,114 @@ oppia.directive('topNavigationBar', [function() {
+            });
+          }
+          currentWindowWidth = windowDimensionsService.getWidth();
+          oppiaDebouncer.debounce(truncateNavbar, 500)();
```
Re the debouncer -- great catch! I took a look, and I think there are a couple of errors in the oppiaDebouncer.debounce():
- Can you try removing the "last > 0" part of the if statement?
- Could you move the timeout, context, args, timestamp and result vars above the general return statement for the factory (so that they become, essentially, private vars in the factory)? Previously they were local variables, and I think that that was what was causing the problem, because the timeout would always be empty.

That said, we could have problems if there are multiple debouncers in a page (and in fact this could happen in the library page since we have the search debouncer). It seems like Angular has its own [debouncer](https://docs.angularjs.org/api/ng/directive/ngModelOptions); do you think we could use that for the search box and reserve this one for the navbar resizing? (We should put a big warning on the debouncer to say that this is only used for the navbar, or move the logic into the TopNavigationBarDirective if it's not being used elsewhere.)

Oishikatta replied:
Making those changes worked, thank you.

Should I add a commit to use Angular's debouncer in the library here or make a new issue and/or PR for that?

I'm not sure what would be best as far as moving it to TopNavigationBarDirective or leaving it in app.js. I had initially thought the factory pattern worked like the constructor pattern, i.e. each use would have its own private vars.

One solution I can do for leaving it in app.js (in case of other future uses) could be changing it to return a constructor function, like this: http://stackoverflow.com/q/18072683

Oishikatta commented:
@seanlip 

I found an alternative approach that I think would be more reliable and simpler to maintain, can I get your input on going this direction instead? My last push implements this.

The idea is to use inline-block instead of float. With that change, there is no need/reason to do any measuring of widths in javascript. The truncateNavbar code is then simplified to basically check the height for overflow and hide elements until there is no longer overflow.

I think part of the reason float was originally used was to support IE8 - but since Oppia doesn't support IE8, that's probably not a concern now.

One of the benefits of this is for example on the library/search pages, there would no longer be a need for different cutoff widths: https://github.com/oppia/oppia/pull/2927/files#diff-1f2b94e630d70b985876b243cbf12f2eL465 Instead, the navbar would just show the elements that are able to fit like on the rest of the pages.

seanlip commented:
Hi @Oishikatta -- thanks, this sounds very promising! I can't take a detailed look now, but I expect the new direction would be good if the UI works well. @jaredsilver, can you confirm this?

Thanks!

jaredsilver commented:
This seems to work as expected from a UI perspective -- nice work, @Oishikatta!

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,91 @@ oppia.directive('topNavigationBar', [function() {
+         */
+        var checkIfI18NCompleted = function() {
+          var i18nCompleted = true;
+          document.querySelectorAll('.oppia-navbar-tabs a[translate], ' +
```
I think it would be cleaner to just add a single class (.oppia-navbar-tab?) to each "navbar tab" and then query for elements with that class.

Oishikatta replied:
Added.

* .oppia-navbar-tab was already in use, so used .oppia-navbar-tab-content

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,91 @@ oppia.directive('topNavigationBar', [function() {
+
+        /**
+         * Checks if window is >768px and i18n is completed, then checks
+         * for overflow. If overflow is detected hides the least important
```
If overflow is detected, hides the least ...

(add comma)

Oishikatta replied:
Added.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,91 @@ oppia.directive('topNavigationBar', [function() {
+        /**
+         * Checks if window is >768px and i18n is completed, then checks
+         * for overflow. If overflow is detected hides the least important
+         * tab and then calls itself again after a 50ms delay.
```
Do we still have a sense of "least important tab"?

Oishikatta replied:
The order is maintained implicitly by the order they're declared in the navElementsVisibilityStatus object. So they're simply declared in order of least to most important, and the browsers follow that when running the for..in loop. ["All modern implementations of ECMAScript iterate through object properties in the order in which they were defined."](http://ejohn.org/blog/javascript-in-chrome/)

I think I should add this information in a comment with the object definition.

seanlip replied:
Hm, yeah -- I think that would be a good idea, though it'd be nice not to rely on not-in-the-spec behavior.

An alternative way is to have an additional array that contains the keys in the right order, and add a karma test that checks that the elements of the array match exactly the keys in the object.

Oishikatta replied:
Added an array to explicitly define the order. I'm not sure about karma tests though - where would I put one for the topNavigationBar directive?

seanlip replied:
Oh yeah -- it's a directive. Directives are a bit annoying to karma-test, actually, so let's not do it. Just add a comment along the lines of: "NOTE TO DEVELOPERS: this array must be kept in sync with the above object."

seanlip replied:
Actually, scrap that. I thought of a better way: define the array, and then _derive_ the object from it using a for loop. That way, there's only one source of truth. I think that's much cleaner.

Oishikatta replied:
That is much nicer, thanks. Added.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,91 @@ oppia.directive('topNavigationBar', [function() {
+        var truncateNavbar = function() {
+          // If the window is narrow, the standard nav tabs are not shown.
+          if (windowDimensionsService.isWindowNarrow()) {
+            return false;
```
What's the point of the return value -- is it ever used?

Oishikatta replied:
Return value wasn't used, removed it.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,91 @@ oppia.directive('topNavigationBar', [function() {
+            return false;
+          }
+
+          // Measured non-overflowed navbar height under 60px via inspector.
```
Sorry, I don't understand this comment. Can you explain/elaborate?

Oishikatta replied:
What I did to get "60px" for the overflow height was to use the inspector in Chrome to measure the normal height of the navbar when not overflowed (56px) and round up. That value is hardcoded, i.e. the function can only detect that the height of the navbar is larger than it was when I measured it.

seanlip replied:
OK, please add this elaboration in the code about how the 60 was obtained. I think it is useful.

Oishikatta replied:
Added.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -95,13 +95,91 @@ oppia.directive('topNavigationBar', [function() {
+        var truncateNavbarDebounced =
+          oppiaDebouncer.debounce(truncateNavbar, 500);
+
+        // For Chrome, timeout 0 appears to run after i18n.
```
Not totally sure what this comment means -- do you mean something like: "In Chrome, this is needed in order to force truncateNavbar to only run after i18n completes."?

Oishikatta replied:
What I mean here is that in Chrome, a timeout of 0 appears to run after i18n completes - but in other browsers (Firefox, Edge), a timeout of 0 will still run before i18n. I used a timeout of 0 since it works for at least one browser, and in others the function will check that i18n isn't done and retry after 100ms.

Should I add "For other browsers, the function will wait for i18n."?

seanlip replied:
Yes, please do (and feel free to err on the more verbose side). All this info is useful!

Oishikatta replied:
Added more verbose comment.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2500,27 +2500,32 @@ div#ng-curtain {
+  vertical-align: top;
+}
+.oppia-navbar-clickable-dropdown:hover ul.dropdown-menu, .oppia-clickable-navbar-element:hover {
+  /* Should this be changed at line 852? */
```
I don't know. Maybe? Not 100% sure why this needs to happen specifically on 'hover' though. But if it works, I think that's fine; it'll take a separate project to actually clean up the navbar CSS properly.

Oishikatta replied:
Removed comment since it works.

Oishikatta commented:
I haven't finished testing on different pages yet, I'm working through a copy of the table from spec and retesting. I'll ask for another look once I'm done with that.

Specifically on the learner view, I'm having difficulty reproducing the icon issue shown in the upper right. What browser/OS is showing it? Regarding the unused space it comes from this (merged but not yet in release) PR: https://github.com/oppia/oppia/pull/2888/files#diff-990365d0ec9c4d50c0da512f42dc8d32R48

The easiest way I could reduce the amount of unused space would be by adding more breakpoints. Other than that I could try adding to the truncateNavbar function something like an if (is exploration player and below 768px) then set the max-width of the exploration-header to the available space.

seanlip commented:
Thanks, @Oishikatta! Just a note -- you may want to resolve the merge conflicts that seem to have appeared. I'm kind of having trouble checking out your branch (I'm trying to retest the upper-left thing). I suspect I was on Chromium v43.0.2357.130 on Ubuntu but I'm not 100% sure.

For the unused space issue, maybe you could file an issue for that, and let's defer that one? Let's try and get this PR in, it's been around for a while :)

seanlip commented:
Oh yeah, I think it still does happen on Chromium. When you try it out, make the screen as narrow as possible and look at the upper left (not the upper right).

Oishikatta commented:
@seanlip I've completed testing on all the pages from the summary table at each screen size, it's ready for another look now. Merge conflict is resolved.

I fixed the issue with the exploration player logo on small screens. 

The only remaining area I found where the navbar doesn't match spec is the Library, in that the category/language option breakpoints are set differently than spec (940, 1275 vs 1024, 1440): https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/library/search_bar_directive.html#L120

They work fine as they are - since they're lower than spec there is less unused space. They could even be lowered to 768px and have the navbar show whatever elements fit. But it would also be simple to change them to match spec.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -94,24 +95,116 @@ oppia.directive('topNavigationBar', [function() {
+          $scope.toggleSidebar = function() {
+            SidebarStatusService.toggleSidebar();
+          };
+          // If the window is resized larger, try displaying the hidden elements
```
nit: add full stop at end.

Oishikatta replied:
Added.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -94,24 +95,116 @@ oppia.directive('topNavigationBar', [function() {
+        };
+        // The order of the elements in this array specifies the order in which
+        // they will be hidden.
+        navElementsOrder = ['I18N_TOPNAV_DONATE', 'I18N_TOPNAV_ABOUT',
```
Declare as constant: NAV_ELEMENTS_ORDER

Oishikatta replied:
Changed to NAV_ELEMENTS_ORDER.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -94,24 +95,116 @@ oppia.directive('topNavigationBar', [function() {
-            angular.element(document.body).addClass('oppia-stop-scroll');
-          } else {
-            angular.element(document.body).removeClass('oppia-stop-scroll');
+          $scope.isSidebarShown = function() {
```
I'm not quite sure why all these helper functions are being defined _inside_ the resize hook?

Oishikatta replied:
This was supposed to be part of the merge, but I did it incorrectly. I compared to https://github.com/oppia/oppia/commit/b9c43c93aaee18b59ec18774381112fe6f998846 and I think I've corrected the merge.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -94,24 +95,116 @@ oppia.directive('topNavigationBar', [function() {
+        // check if i18n is complete and set a new timeout if it is not. Since
+        // a timeout of 0 works for at least one browser, it is used here.
+        $timeout(truncateNavbar, 0);
+        $scope.toggleSidebar = SidebarStatusService.toggleSidebar;
```
This seems to be redefined above (line 124). Is one redundant?

Oishikatta replied:
This was a mistake with the merge, I've corrected it.

seanlip commented on core/templates/dev/head/pages/library/SearchBarDirective.js:
```diff
@@ -205,6 +205,7 @@ oppia.directive('searchBar', [function() {
             }
 
             refreshSearchBarLabels();
+            $rootScope.$broadcast('searchBarLoaded', true);
```
Why is this needed (what broke before)?

Oishikatta replied:
I noticed that sometimes on the library/search page, the custom nav elements from the SearchBarDirective would load/show up after the function had already run, causing the overflow not to be handled. Sending the event from there makes sure to run the function again late enough to work.

seanlip replied:
Ah, OK. I think this could do with an explanatory comment.

Oishikatta replied:
Added.

Oishikatta commented:
It works properly when I tested with various languages both via the dropdown and preferences.

> That said, is making everything 768 px a workable and easy-to-implement solution? If so, and if it doesn't impair the user experience, maybe we should do that, so that we don't have so many numbers floating around the place. What do you think?

Making the breakpoints for the library category/language options 768px (or anything else) is trivial, only involves updating the two lines of CSS.

I found that there would be overflow from 769~774px in some languages (e.g. Dutch) even with all buttons hidden.

I would suggest changing the breakpoints to 768px (categories) and 1024px (languages), but the caveat is that would cause the 'Donate' button to be hidden from about 768~1100px.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -94,12 +95,34 @@ oppia.directive('topNavigationBar', [function() {
+        var currentWindowWidth = windowDimensionsService.getWidth();
+        $scope.navElementsVisibilityStatus = {};
+        // The order of the elements in this array specifies the order in which
+        // they will be hidden.
```
Add a note: "XXX elements will be hidden first." (where XXX == Earlier/Later)

Oishikatta replied:
Added.

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -94,12 +95,34 @@ oppia.directive('topNavigationBar', [function() {
+        $scope.navElementsVisibilityStatus = {};
+        // The order of the elements in this array specifies the order in which
+        // they will be hidden.
+        NAV_ELEMENTS_ORDER = ['I18N_TOPNAV_DONATE', 'I18N_TOPNAV_ABOUT',
```
missing var?

Oishikatta replied:
Added.

seanlip commented:
Hm, weird. I have restarted the build. Let's see if they persist.

(If they do, @Oishikatta, can you merge develop into this branch again, just to make sure everything's fully up to date? Thanks!)

Oishikatta commented:
> Can you file a separate issue to investigate overflow in long translations? We may be able to substitute them with shorter ones.

There would only be overflow if the breakpoints were set to 768/768; since they're now 768/1024, there is no overflow.

> I think 768px and 1024px sounds good. I'm fine with the 'Donate' button being hidden where necessary. (I assume this is just on the library page?)

Set breakpoints for category/language inputs on library/search to 768/1024px. Right, it's only on the library/search pages.

I've merged current develop in, but some of the tests still fail. I'm not familiar with them but I am looking into it.
